### PR TITLE
Itermaggedon: update everything for 0.7-only

### LINKED ISSUE #29: Exponential time complexity for recursive chains with increasing NTuple length
Hi all,

there is some strange exponential time and memory behavior that I don't understand:
I have a tree with some additional data (`Vector{Foo}`) on every node, and eventually I want to iterate through all the `Foo` objects without unnecessary allocations. At first I thought I ran into an infinite loop, because CPU got stuck at 100% and nothing else would happen, but I was able to track down the problem to the following minimal example. I'll just use a linear list instead of a wider tree here, and run `length()` instead of collecting items. That is enough to show the exponential behavior:
```julia
versioninfo()

using IterTools

struct Foo end

"""
Creates a Chain{Tuple{Chain{Tuple{ ... NTuple{n,Vector{Foo}} ... }}}} of depth `d`,
with an NTuple of length `n` on the lowest level, containing empty Vectors of `Foo`.
"""
function g(d::Int, n::Int)
  if d == 0
    return [Foo()]
  elseif d == 1
    return chain([g(d - 1, n) for i in 1:n]...)
  else
    return chain(g(d - 1, n))
  end
end

for i in 1:19
  @time length(g(5, i))
end
```

Result for the above example (d=5, n=19):
```julia
Julia Version 0.6.2
Commit d386e40c17 (2017-12-13 18:08 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-5820K CPU @ 3.30GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, haswell)

  0.184802 seconds (123.40 k allocations: 6.813 MiB, 26.74% gc time)
  0.115444 seconds (123.95 k allocations: 6.828 MiB)
  0.115224 seconds (129.93 k allocations: 7.195 MiB)
  0.175451 seconds (136.42 k allocations: 7.585 MiB, 31.63% gc time)
  0.120656 seconds (143.94 k allocations: 8.109 MiB)
  0.124801 seconds (153.71 k allocations: 8.956 MiB)
  0.176149 seconds (168.00 k allocations: 10.415 MiB, 28.02% gc time)
  0.141948 seconds (191.84 k allocations: 13.889 MiB, 2.80% gc time)
  0.144715 seconds (235.74 k allocations: 20.443 MiB)
  0.167659 seconds (321.68 k allocations: 37.108 MiB, 3.56% gc time)
  0.208135 seconds (495.73 k allocations: 69.814 MiB, 3.25% gc time)
  0.300323 seconds (854.18 k allocations: 138.819 MiB, 3.42% gc time)
  0.487694 seconds (1.60 M allocations: 307.199 MiB, 3.67% gc time)
  0.874714 seconds (3.14 M allocations: 639.010 MiB, 3.37% gc time)
  1.708150 seconds (6.36 M allocations: 1.311 GiB, 3.37% gc time)
  3.567857 seconds (13.05 M allocations: 3.163 GiB, 3.62% gc time)
  7.628431 seconds (26.95 M allocations: 6.670 GiB, 3.52% gc time)
 16.260810 seconds (55.80 M allocations: 14.041 GiB, 3.78% gc time)
 34.626086 seconds (115.57 M allocations: 29.456 GiB, 4.18% gc time)
```

If I run the same loop again, the times are suddenly what I would've expected in the first place:
```julia
  0.000087 seconds (27 allocations: 992 bytes)
  0.000060 seconds (29 allocations: 1.094 KiB)
  0.000045 seconds (31 allocations: 1.219 KiB)
  0.000047 seconds (33 allocations: 1.344 KiB)
  0.000049 seconds (35 allocations: 1.469 KiB)
  0.000047 seconds (37 allocations: 1.594 KiB)
  0.000042 seconds (39 allocations: 1.719 KiB)
  0.000041 seconds (41 allocations: 1.844 KiB)
  0.000045 seconds (43 allocations: 1.969 KiB)
  0.000042 seconds (45 allocations: 2.094 KiB)
  0.000040 seconds (47 allocations: 2.219 KiB)
  0.000041 seconds (49 allocations: 2.344 KiB)
  0.000040 seconds (51 allocations: 2.469 KiB)
  0.000039 seconds (53 allocations: 2.594 KiB)
  0.000042 seconds (55 allocations: 2.719 KiB)
  0.000046 seconds (57 allocations: 2.844 KiB)
  0.000048 seconds (59 allocations: 2.969 KiB)
  0.000044 seconds (61 allocations: 3.094 KiB)
  0.000045 seconds (63 allocations: 3.219 KiB)
```

I guess my question is: Why is that? And how can I fix it efficiently? I could change `chain` to `vcat`, but I was hoping to avoid any unnecessary copies/allocations.

### LINKED ISSUE #33: Doc - difference between Base.Iterators.product and IterTools.product
Hello,

Maybe we should add in doc an example showing difference between `Base.Iterators.product` and `IterTools.product`

```julia
julia> xss = 1:3, 5:7, 10:11
(1:3, 5:7, 10:11)

julia> collect(Base.Iterators.product(xss...))
3×3×2 Array{Tuple{Int64,Int64,Int64},3}:
[:, :, 1] =
 (1, 5, 10)  (1, 6, 10)  (1, 7, 10)
 (2, 5, 10)  (2, 6, 10)  (2, 7, 10)
 (3, 5, 10)  (3, 6, 10)  (3, 7, 10)

[:, :, 2] =
 (1, 5, 11)  (1, 6, 11)  (1, 7, 11)
 (2, 5, 11)  (2, 6, 11)  (2, 7, 11)
 (3, 5, 11)  (3, 6, 11)  (3, 7, 11)

julia> collect(IterTools.product(xss...))
18-element Array{Tuple{Int64,Int64,Int64},1}:
 (1, 5, 10)
 (2, 5, 10)
 (3, 5, 10)
 (1, 6, 10)
 (2, 6, 10)
 (3, 6, 10)
 (1, 7, 10)
 (2, 7, 10)
 (3, 7, 10)
 (1, 5, 11)
 (2, 5, 11)
 (3, 5, 11)
 (1, 6, 11)
 (2, 6, 11)
 (3, 6, 11)
 (1, 7, 11)
 (2, 7, 11)
 (3, 7, 11)
```

Kind regards

### LINKED ISSUE #34: New Iteration Protocol
**Keno**:
> The recommendation is to either define iteration conditionally (if VERSION, then define one or the other) or just drop support for 0.6

I propose we have one file for 0.6 iterators and one for 0.7 iterators. I doubt there will be much shared code. Then once 1.0 is out we can just delete the 0.6 stuff.

I will probably start working on this soon so please post if you have any thoughts.

### PR DESCRIPTION
**Pull request by iamed2**
Let's shave some bikes and paint some yaks!

1. This is already a pretty major change. Should we just ditch `chain` and `product` rather than deprecating them?
2. Are we okay getting rid of the old `chain` and `product`? They were worse than the Base versions for a minor benefit (length worked and eltype worked better).
3. How do we feel about the `imap` change?
4. What should we call `@something`? `something` is now a function so we should probably not call it that.

Note: 0.7 tests will error until the 0.7-alpha binaries are posted. Nightly tests should pass.

Closes #34 
Closes #33 (if we deprecate product) and closes #29 (if we deprecate chain)

### CODE REVIEW

File: src/IterTools.jl [hunk 1]
```diff
@@ -156,152 +143,20 @@ julia> collect(repeatedly(t, 5))
...
-        end
-
-        js[i] = start(it.xss[i])
-        vs[i], js[i] = next(it.xss[i], js[i])
-    end
-    ans, (js, nothing)
-end
-
-done(it::Product, state) = state[2] === nothing
+@deprecate chain(xss...) Iterators.flatten(xss)
```
Comment (ararslan) [inline src/IterTools.jl]: Don't these behave differently when `xss` is empty?

Comment (iamed2) [reply src/IterTools.jl]: Yes, flatten does this:

```julia
julia> collect(Iterators.flatten(()))
ERROR: MethodError: Base.IteratorEltype(::Type{Union{}}) is ambiguous. Candidates:
  Base.IteratorEltype(::Type{#s59} where #s59<:Base.Iterators.Repeated) in Base.Iterators at iterators.jl:726
  Base.IteratorEltype(::Type{#s59} where #s59<:Base.Iterators.PartitionIterator{T}) where T in Base.Iterators at iterators.jl:941
  Base.IteratorEltype(::Type{#s624} where #s624<:Base.Broadcast.Broadcasted) in Base.Broadcast at broadcast.jl:260
  Base.IteratorEltype(::Type{#s59} where #s59<:(Base.Iterators.Rest{I,S} where S)) where I in Base.Iterators at iterators.jl:502
Possible fix, define
  Base.IteratorEltype(::Type{Union{}})
```

which is...a bug?

Comment (iamed2) [reply src/IterTools.jl]: In any case I don't expect these to be direct replacements, which is why I was considering not having deprecations. `IterTools.product` and `Iterators.product` had a different iteration order, shape, and 0-element behaviour, but I don't think anyone really needs the IterTools behaviour.

**Comment by iamed2** [comment]
Looks like deprecating `chain` will break Gadfly, so we should see if they can use `Iterators.flatten` instead.

**Comment by iamed2** [comment]
Compose uses both `chain` and `product`, and may depend on the ordering for `product`. We should attempt to update that one first.

**Comment by iamed2** [comment]
Text.jl needs a simple fix to use `flatten` instead of `chain`.

**Comment by iamed2** [comment]
PhyloNetworks.jl may depend on the ordering of `product`

**Comment by ararslan** [comment]
Nice analysis of packages! Is it the empty case for `chain` that will break Gadfly when using `flatten`?

**Comment by iamed2** [comment]
BasisMatrices.jl uses `product` and depends on order but we can switch it to `Iterators.product` order by swapping some indices

**Comment by iamed2** [comment]
@ararslan In Gadfly they actually specialize a method on `IterTools.chain`, so that's the main thing that needs to be fixed

**Comment by iamed2** [comment]
I _think_ BioSequences.jl doesn't rely on the ordering of `product` but it's probably good to check.

**Comment by iamed2** [comment]
## Package Updates

- [ ] Gadfly
  - [x] [PR](https://github.com/GiovineItalia/Gadfly.jl/pull/1159)
  - [ ] Merged
  - [ ] Tagged
- [ ] Compose
  - [x] [PR](https://github.com/GiovineItalia/Compose.jl/pull/285)
  - [ ] Merged
  - [ ] Tagged
- [ ] Text
  - [ ] PR
  - [ ] Merged
  - [ ] Tagged
- [ ] PhyloNetworks
  - [ ] PR
  - [ ] Merged
  - [ ] Tagged
- [ ] BasisMatrices
  - [ ] PR
  - [ ] Merged
  - [ ] Tagged
- [ ] BioSequences
  - [ ] PR
  - [ ] Merged
  - [ ] Tagged

**Comment by ararslan** [comment]
@iamed2 Is this ready to merge?

**Comment by iamed2** [comment]
Just realized I should give this a Project file

**Comment by ararslan** [comment]
Thanks for your amazing work here!

**Pull request merged**